include/deal.II/non_matching/fe_values.h

```diff
@@ -0,0 +1,511 @@
+   * points. This can, for example, happen if the cut is of floating point size.
+   * Since the FEValues-like objects aren't allowed to contain 0 points, the
+   * object that get_inside/outside/surface_fe_values() returns is wrapped in a
+   * boost::optional. This requires us to check if the returned FEValues-like
+   * object contains a value before we use it:
+   * @code
+   * NonMatching::FEValues<dim> fe_values(...);
+   *
+   * for (const auto &cell : dof_handler.active_cell_iterators())
+   *  {
+   *    fe_values.reinit(cell);
+   *
+   *    const boost::optional<const FEValues<dim> &> fe_values_inside =
+   *      fe_values.get_inside_fe_values();
+   *
+   *    if (fe_values_inside)
+   *      {
+   *        // Assemble locally
+   *        for (const unsigned int q_index :
+   *             fe_values_inside->quadrature_point_indices())
+   *          {
+   *            // ...
+   *          }
+   *      }
+   *  }
+   * @endcode
+   *
+   * Of course, it is somewhat expensive to generate the immersed quadrature
+   * rules and create FEValues objects with the generated quadratures. To reduce
+   * the amount of work, the reinit() function of this class uses the
+   * MeshClassifier passed to the constructor to check how the incoming cell
+   * relates to the level set function. It only generates the immersed
+   * quadrature rules if the cell is intersected. If the cell is completely
+   * inside or outside it returns a cached FEValues object created with a
+   * quadrature over the reference cell: $[0, 1]^{dim}$.
+   */
+  template <int dim>
+  class FEValues
+  {
+  public:
+    using AdditionalData = typename QuadratureGenerator<dim>::AdditionalData;
+
+    /**
+     * Constructor.
+     *
+     * @param fe_collection Collection of FiniteElements to be used.
+     * @param quadrature 1-dimensional quadrature rule used to generate the
+     * immersed quadrature rules. See the QuadratureGenerator class. On the non
+     * intersected elements a tensor product of this quadrature will be used.
+     * @param mesh_classifier Object used to determine when the immersed
+     * quadrature rules need to be generated.
+     * @param region_update_flags Struct storing UpdateFlags for the
+     * inside/outside/surface region of the cell.
+     * @param dof_handler The DoFHandler associated with the discrete level set
+     * function.
+     * @param level_set The degrees of freedom of the discrete level set function.
+     * @param additional_data Additional data passed to the QuadratureGenerator
+     * class.
+     *
+     * @note  Pointers to @p fe_collection, @p mesh_classifier, @p dof_handler,
+     * and @p level_set are stored internally, so these need to have a longer life
+     * span than the instance of this class.
+     */
+    template <class VectorType>
+    FEValues(const hp::FECollection<dim> &fe_collection,
+             const Quadrature<1> &        quadrature,
```

Comment:
```suggestion
             const Quadrature<dim> &        quadrature,
```
You could use internally `Quadrature::is_tensor_product()` and `Quadrature::get_tensor_basis`.

Reply:
Yes, but wouldn't that be more unclear? One would first have to assert that the incoming `Quadrature<dim>` is a tensor product and that each `Quadrature<1>` in the basis is the same, otherwise we have to choose which component to build the immersed quadrature rules from.

Reply:
The reason why I would prefer `Quadrature<dim>` is that `FEValues` takes one of these, i.e., there will be probably an instance of this lying around. What opinion have others?

Reply:
I think in >90% of cases one creates a quadrature right at the place of setting up `FEValues`, so in that sense I do not believe that this should be the deciding factor. Indeed, the 1D quadrature formula signals clearly to the user that we need a very specific tensor product, due to the nature in the construction of the cut quadrature formula at this point. For me, the criterion would be whether we can think of an extension of the interface to a different class of algorithms to define cut quadrature formulas and if we can anticipate anyone interested in contributing these in the mid term (say in the next few years). @bergbauer what is your opinion here?

If we go with the dim-dimensional quadrature formula, we need to write more documentation, and we need to find good ways to compare quadrature objects as suggested by @simonsticko - we would need to inspect the `get_tensor_basis` objects. At least it is not hopeless to do so, so I would make it dependent on the possibility for extension.